src/core/atomic.d

```diff
@@ -32,12 +32,62 @@ else
 
 private
 {
-    template HeadUnshared(T)
+    /* Construct a type with a shared tail, and if possible with an unshared
+    head. */
+    template TailShared(T)
     {
-        static if( is( T U : shared(U*) ) )
-            alias shared(U)* HeadUnshared;
+        alias S = shared T;
+
+        static if (is(S U == shared U)) {}
+        else static assert(false);
```

Comment:
> I would specify that it should never be hit in the error message and the reason why

I've added some more text. I may have gone overboard. If it's too verbose now, I'd appreciate a concrete suggestion.